python/ray/air/tests/test_data_batch_conversion.py

```diff
@@ -16,40 +16,72 @@ def test_pandas_pandas():
     input_data = pd.DataFrame({"x": [1, 2, 3]})
     expected_output = input_data
     actual_output = convert_batch_type_to_pandas(input_data)
-    assert expected_output.equals(actual_output)
+    pd.testing.assert_frame_equal(expected_output, actual_output)
+
+    actual_output = convert_pandas_to_batch_type(actual_output, type=DataType.PANDAS)
+    pd.testing.assert_frame_equal(actual_output, input_data)
+
+
+@pytest.mark.parametrize("use_tensor_extension_for_input", [True, False])
+@pytest.mark.parametrize("cast_tensor_columns", [True, False])
+def test_pandas_multi_dim_pandas(cast_tensor_columns, use_tensor_extension_for_input):
+    input_tensor = np.arange(12).reshape((3, 2, 2))
+    input_data = pd.DataFrame(
+        {
+            "x": TensorArray(input_tensor)
+            if use_tensor_extension_for_input
+            else list(input_tensor)
+        }
+    )
+    expected_output = pd.DataFrame(
+        {
+            "x": (
+                list(input_tensor)
+                if cast_tensor_columns or not use_tensor_extension_for_input
+                else TensorArray(input_tensor)
+            )
+        }
+    )
+    actual_output = convert_batch_type_to_pandas(input_data, cast_tensor_columns)
+    pd.testing.assert_frame_equal(expected_output, actual_output)
 
-    assert convert_pandas_to_batch_type(actual_output, type=DataType.PANDAS).equals(
-        input_data
+    actual_output = convert_pandas_to_batch_type(
+        actual_output, type=DataType.PANDAS, cast_tensor_columns=cast_tensor_columns
     )
+    pd.testing.assert_frame_equal(actual_output, input_data)
 
 
-def test_numpy_pandas():
+@pytest.mark.parametrize("cast_tensor_columns", [True, False])
+def test_numpy_pandas(cast_tensor_columns):
     input_data = np.array([1, 2, 3])
-    expected_output = pd.DataFrame({TENSOR_COLUMN_NAME: TensorArray([1, 2, 3])})
-    actual_output = convert_batch_type_to_pandas(input_data)
-    assert expected_output.equals(actual_output)
+    expected_output = pd.DataFrame({TENSOR_COLUMN_NAME: input_data})
+    actual_output = convert_batch_type_to_pandas(input_data, cast_tensor_columns)
+    pd.testing.assert_frame_equal(expected_output, actual_output)
```

Comment:
for my own understanding, does pd.testing.assert_frame_equal work regardless of TensorArray?

Reply:
Yes, it should AFAIK!

Reply:
Ah do you mean comparing between a tensor extension type column and the equivalent opaque object column? I believe that will fail since the column dtypes will different and comparing an ndarray and an ndarray of ndarray pointers will fail in NumPy comparison land.